torch/lib/THD2/FileStore.cpp

```diff
@@ -0,0 +1,270 @@
+    flock(operation);
+  }
+
+  ~Lock() {
+    if (fd_ >= 0) {
+      flock(LOCK_UN);
+    }
+  }
+
+  Lock(const Lock& that) = delete;
+
+  Lock(Lock&& other) noexcept {
+    fd_ = other.fd_;
+    other.fd_ = -1;
+  }
+
+ protected:
+  int fd_;
+
+  void flock(int operation) {
+    while (true) {
+      auto rv = ::flock(fd_, operation);
+      if (rv == -1) {
+        if (errno == EINTR) {
+          continue;
+        }
+        throw std::system_error(errno, std::generic_category(), "flock");
+      }
+      break;
+    }
+  }
+};
+
+class File {
+ public:
+  explicit File(const std::string& path, int flags) {
+    while (true) {
+      fd_ = ::open(path.c_str(), flags, 0644);
+      if (fd_ < 0) {
+        if (errno == EINTR) {
+          continue;
+        }
+        throw std::system_error(
+            errno,
+            std::generic_category(),
+            "Unable to open " + path);
+      }
+      break;
+    }
+  }
+
+  ~File() {
+    ::close(fd_);
+  }
+
+  Lock lockShared() {
+    return Lock(fd_, LOCK_SH);
+  }
+
+  Lock lockExclusive() {
+    return Lock(fd_, LOCK_EX);
+  }
+
+  off_t seek(off_t offset, int whence) {
+    auto rv = lseek(fd_, offset, whence);
+    SYSASSERT(rv);
+    return rv;
+  }
+
+  off_t tell() {
+    auto rv = lseek(fd_, 0, SEEK_CUR);
+    SYSASSERT(rv);
+    return rv;
+  }
+
+  off_t size() {
+    auto pos = tell();
+    auto size = seek(0, SEEK_END);
+    seek(pos, SEEK_SET);
+    return size;
+  }
+
+  void writeString(const std::string& str) {
+    uint32_t len = str.size();
+    SYSCHECK(write(fd_, &len, sizeof(len)));
+    SYSCHECK(write(fd_, str.c_str(), len));
+  }
+
+  void readString(std::string& str) {
+    uint32_t len;
+    SYSCHECK(read(fd_, &len, sizeof(len)));
+    std::vector<char> buf(len);
+    SYSCHECK(read(fd_, buf.data(), len));
+    str.assign(buf.begin(), buf.end());
+  }
+
+  void writeVector(const std::vector<char>& data) {
+    uint32_t len = data.size();
+    SYSCHECK(write(fd_, &len, sizeof(len)));
+    SYSCHECK(write(fd_, data.data(), len));
+  }
+
+  void readVector(std::vector<char>& data) {
+    uint32_t len;
+    SYSCHECK(read(fd_, &len, sizeof(len)));
+    data.resize(len);
+    SYSCHECK(read(fd_, data.data(), len));
+  }
+
+ protected:
+  int fd_;
+};
+
+} // namespace
+
+FileStore::FileStore(const std::string& path)
+    : Store(),
+      path_(path),
+      pos_(0) {
+}
+
+FileStore::~FileStore() {
+}
+
+void FileStore::set(const std::string& key, const std::vector<char>& value) {
+  File file(path_, O_RDWR | O_CREAT);
+  auto lock = file.lockExclusive();
+  file.seek(0, SEEK_END);
+  file.writeString(key);
+  file.writeVector(value);
+}
+
+std::vector<char> FileStore::get(const std::string& key) {
+  while (cache_.count(key) == 0) {
+    File file(path_, O_RDONLY);
+    auto lock = file.lockShared();
+    auto size = file.size();
+    if (size == pos_) {
+      // No new entries; sleep for a bit
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
```

Comment:
Good point. The lock will be released because it goes out of scope, but only after the 10 milliseconds. We should release early so the other process doesn't have to wait for nothing.